python/pyarrow/_dataset.pyx

```diff
@@ -0,0 +1,1110 @@
+        """Create a dataset
+
+        A schema must be passed because most of the data sources' schema is
+        unknown before executing possibly expensive scanning operation, but
+        projecting, filtering, predicate pushduwn requires a well defined
+        schema to work on.
+
+        Parameters
+        ----------
+        data_sources : list of DataSource
+            One or more input data sources
+        schema : Schema
+            A known schema to conform to.
+        """
+        cdef:
+            DataSource source
+            CDataSourceVector sources
+            CResult[CDatasetPtr] result
+            shared_ptr[CSchema] sp_schema
+
+        for source in data_sources:
+            sources.push_back(source.wrapped)
+
+        result = CDataset.Make(sources, pyarrow_unwrap_schema(schema))
+
+        self.init(GetResultValue(result))
+
+    cdef void init(self, const shared_ptr[CDataset]& sp):
+        self.wrapped = sp
+        self.dataset = sp.get()
+
+    cdef inline shared_ptr[CDataset] unwrap(self):
+        return self.wrapped
+
+    def new_scan(self, ScanContext context=None):
+        """Begin to build a new Scan operation against this Dataset."""
+        cdef CResult[shared_ptr[CScannerBuilder]] result
+        if context is None:
+            result = self.dataset.NewScan()
+        else:
+            result = self.dataset.NewScanWithContext(context.unwrap())
+        return ScannerBuilder.wrap(GetResultValue(result))
+
+    @property
+    def sources(self):
+        cdef vector[shared_ptr[CDataSource]] sources = self.dataset.sources()
+        return [DataSource.wrap(source) for source in sources]
+
+    @property
+    def schema(self):
+        return pyarrow_wrap_schema(self.dataset.schema())
+
+
+cdef class ScanContext:
+
+    cdef:
+        shared_ptr[CScanContext] wrapped
+        CScanContext *context
+
+    def __init__(self, MemoryPool memory_pool=None):
+        cdef shared_ptr[CScanContext] context = make_shared[CScanContext]()
+        self.init(context)
+        if memory_pool is not None:
+            self.memory_pool = memory_pool
+
+    cdef init(self, shared_ptr[CScanContext]& sp):
+        self.wrapped = sp
+        self.context = sp.get()
+
+    @staticmethod
+    cdef ScanContext wrap(shared_ptr[CScanContext]& sp):
+        cdef ScanContext self = ScanContext.__new__(ScanContext)
+        self.init(sp)
+        return self
+
+    cdef inline shared_ptr[CScanContext] unwrap(self):
+        return self.wrapped
+
+    @property
+    def memory_pool(self):
+        # TODO(kszucs): there's no function to box/wrap memory pool
+        cdef:
+            MemoryPool pool = MemoryPool.__new__(MemoryPool)
+        # this might be unsafe because of the raw pointers
+        pool.init(self.context.pool)
+        return pool
+
+    @memory_pool.setter
+    def memory_pool(self, MemoryPool pool):
+        self.context.pool = maybe_unbox_memory_pool(pool)
+
+
+cdef class ScanTask:
```

Comment:
There was some discussion about having Scanner directly yield record batches (`scanner.scan() -> Iterator[RecordBatch]`), but not sure any more if there was some conclusion. Having the scan tasks allows a user to do custom (within process) scheduling?

Reply:
Ah, missed your answer in another inline code comment thread about this.

Reply:
Switched the API to return `Iterator[RecordBatch]`

Reply:
Switched the API to return `Iterator[RecordBatch]`